### DISCUSSION TITLE
Connection Pooling support in faraday adapters

### DISCUSSION THREAD

**Discussion by MaryamAdnan3** [lostisland/faraday#1421] [Help] [answered]
I was looking for connection pooling support in different faraday adapters and found #1024 and #1006 which show work in progress. #1006 mentions that certain adapters like Net-Http and Excon are incompatible by design to support connection pooling. Are there any other options available to add support for connection pooling without depending on the underlying support of adapters? There are certain gems that add generic support for connection pooling regardless of the underlying Ruby client e.g https://github.com/mperham/connection_pool. Can this gem help in adding connection pooling support in adapters like net-http, excon etc?
Code Snippet:
https://gist.github.com/MaryamAdnan3/7c08b0cda0b5e8486fe2079911da094f

**Comment by iMacTia** [comment, marked as answer]
Hey @MaryamAdnan3 👋, [`connection_pool`](https://github.com/mperham/connection_pool) is exactly the gem I used in #1006, but that has nothing to do with HTTP connections. That gem is simply providing you with mutexed access to shared resources. This is obviously useful for managing connections across threads/fibers, but you could technically put whatever you want in there 😄.

In fact, in your [code snippet](https://gist.github.com/MaryamAdnan3/7c08b0cda0b5e8486fe2079911da094f) you're actually storing `Faraday::Connection` objects in the pool. This actually works, but I doubt is achieving the desired behaviour.
For example, if your `Faraday.default_adapter` is `net_http`, that means that when you check in a connection and use it to make a request, you're still creating a new `Net::HTTP` object, which will open a new socket/connection anyway.

That is where the `connection_pool` really comes useful, by keeping your connections open and reusing them across requests, because that's the "expensive" part. But that's not what you get with your code unless you're using an adapter that already keeps the connection open (e.g. `net_http_persistent`).

So in short, your code works and doesn't really need any change to Faraday to support it 👍, but it only makes sense under certain circumstances (i.e. the underlying adapter keeps the connection open).
It's still a perfectly valid solution if you know what adapter you're using.

**Reply by MaryamAdnan3** [reply]
Thanks, @iMacTia for the explanation.
